# Counting and probability

Counting is essential to probability because we need to be able to count how many outcomes a specific sample space contains and also how many outcomes build up to a specific event so we can calculate its probability. 

Let's think about an example: 

> You love sneakers, but not all of them. You have 1 pair in color Red, Blue, Green and Yellow. And also for each color, you have 1 of each brand you like: Nike, Adidas and Puma. Let's say that you want to randomly pick a sneaker from your collection without looking at it. Every pair of sneakers has the same probability of being drawn. What's the probability that you're gonna pick a pair of color yellow from Adidas?

To start we need to understand how many pairs of sneakers you have available in your collection, i. e., what is the sample space $ S $?

It's known that you have one pair of each color x brand.

| **Color**  | **Brand** | **Brand** | **Brand** |
|:----------:|:---------:|:---------:|:---------:|
| **Red**    | Nike      | Adidas    | Puma      |
| **Blue**   | Nike      | Adidas    | Puma      |
| **Green**  | Nike      | Adidas    | Puma      |
| **Yellow** | Nike      | Adidas    | Puma      |

For each color you have one pair of each brand, i. e. for each of the 4 colors there are 3 different pairs:

# $$ S = 3 \times 4 = 12 $$

The event we are interested in is:

$$A = \text{A yellow pair from Adidas} $$

The number of outcomes (the number of pairs) that favor event A is just 1:

# $$ P(A) = \frac{1}{12} = 0.0833 $$

In [1]:
colors = ["Red", "Blue", "Green", "Yellow"]
brands = ["Nike", "Adidas", "Puma"]
sample_space = 0
favourable_outcomes = 0

# for each color
for color in colors:
    # for each brand
    for brand in brands:
        if color == "Yellow" and brand == "Adidas":
            favourable_outcomes += 1
        sample_space += 1

probability = favourable_outcomes / sample_space

print(f"There are {sample_space} pairs of sneakers.")
print(f"Only {favourable_outcomes} is Yellow from Adidas.")
print(f"The probability of picking a yellow pair from Adidas is: {probability}")

There are 12 pairs of sneakers.
Only 1 is Yellow from Adidas.
The probability of picking a yellow pair from Adidas is: 0.08333333333333333


## Using probability: the multiplicative law

The event we're interested in is a yellow pair (A) AND from Adidas(B).

# $$ P(A \cap B) = P(A) P(B) $$

There are 4 colors, 1 is yellow so:


# $$ P(A) = \frac{1}{4} = 0.25 $$

There are 3 brands, 1 is Adidas so:

# $$ P(B) = \frac{1}{3} = 0.3333 $$

The probability that a pair that was randomly drawn is yellow AND Adidas is:

# $$ P(A \cap B) = 0.25 \times 0.3333 = 0.08333 $$



## Probability tree

One other way that helps with counting is by using a probability tree:

<p align="center">
  <img src="./imgs/probability_tree.png" alt="Probability tree"/>
</p>

You can see that from the 12 possible outcomes from this tree (sample space) only 1 outcome is favorable to the event of a yellow pair from Adidas.

You can see that on each branch I typed the probability of that specific event at the given tree level to happen. If you multiply the probabilities down the branches (in purple color) you'll get the same result as used with the formula above.

Important to note that since the events are mutually exclusive and collectively exhaustive the sum of their probabilities should be 1.

In [2]:
0.0833333333 * 12

0.9999999996000001

For simple problems like this, it's easier to reason about the sample space and events, but if problems are more complex it becomes difficult to count by hand, we'll need some way to abstract this idea so the calculation is a bit easier and faster. On to the next notebook.